[WebGPU] GPURenderPassEncoder.drawIndexed takes significant time in GPU process

mwyrzykowski opened the pull request:
#### 2b68654983dbb899811e57e7280f0c3b429249b6
<pre>
[WebGPU] GPURenderPassEncoder.drawIndexed takes significant time in GPU process
<a href="https://bugs.webkit.org/show_bug.cgi?id=289480">https://bugs.webkit.org/show_bug.cgi?id=289480</a>
<a href="https://rdar.apple.com/146676364">rdar://146676364</a>

Reviewed by Tadeu Zagallo.

executePreDrawCommands was often performing identical validation repeatedely,
leading to to slower framerates than Chrome on some Unity content.

Reduce overhead via caching.

Improves performance on M1 to &gt;60fps in given test cases.

* Source/WebCore/Modules/WebGPU/Implementation/WebGPUComputePassEncoderImpl.cpp:
(WebCore::WebGPU::ComputePassEncoderImpl::setBindGroup):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPURenderPassEncoderImpl.cpp:
(WebCore::WebGPU::RenderPassEncoderImpl::setBindGroup):
* Source/WebGPU/WebGPU/BindGroup.h:
(WebGPU::BindGroup::create):
(WebGPU::BindGroup::vertexArgumentBufferUniqueId const):
(WebGPU::BindGroup::fragmentArgumentBufferUniqueId const):
(WebGPU::BindGroup::computeArgumentBufferUniqueId const):
(WebGPU::BindGroup::uniqueId const):
* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::Device::createBindGroup):
(WebGPU::BindGroup::BindGroup):
(WebGPU::BindGroup::setLabel):
(WebGPU::BindGroup::makeSubmitInvalid const):
(WebGPU::makePipelineBindGroupKey):
(WebGPU::BindGroup::validatedSuccessfully const):
(WebGPU::BindGroup::previouslyValidatedBindGroup const):
* Source/WebGPU/WebGPU/BindGroupLayout.h:
(WebGPU::BindGroupLayout::create):
(WebGPU::BindGroupLayout::uniqueId const):
* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::Device::createBindGroupLayout):
(WebGPU::BindGroupLayout::errorValidatingDynamicOffsets const):
(WebGPU::makeBindGroupLayoutPairIdentifier):
(WebGPU::Device::isCachedCompatibile const):
(WebGPU::Device::setCachedCompatibile const):
(WebGPU::Device::removeCachedBindGroupLayout const):
(WebGPU::BindGroupLayout::errorValidatingBindGroupCompatibility const):
(WebGPU::BindGroupLayout::BindGroupLayout):
(WebGPU::BindGroupLayout::~BindGroupLayout):
* Source/WebGPU/WebGPU/BindableResource.h:
(WebGPU::BufferWithUniqueId::operator id&lt;MTLBuffer&gt; const):
(WebGPU::BufferWithUniqueId::operator uint64_t const):
* Source/WebGPU/WebGPU/Buffer.h:
(WebGPU::Buffer::create):
(WebGPU::Buffer::uniqueId const):
(WebGPU::Buffer::indirectIndexedBufferUniqueId const):
(WebGPU::Buffer::indirectBufferUniqueId const):
* Source/WebGPU/WebGPU/Buffer.mm:
(WebGPU::Device::safeCreateBuffer const):
(WebGPU::Device::createBuffer):
(WebGPU::Buffer::Buffer):
(WebGPU::Buffer::destroy):
(WebGPU::Buffer::getMappedRange):
(WebGPU::Buffer::unmap):
(WebGPU::Buffer::setLabel):
(WebGPU::Buffer::currentSize const):
(WebGPU::Buffer::takeSlowIndexValidationPath):
(WebGPU::Buffer::takeSlowIndirectIndexValidationPath):
(WebGPU::Buffer::takeSlowIndirectValidationPath):
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::encoderIsCurrent const):
* Source/WebGPU/WebGPU/ComputePassEncoder.h:
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::ComputePassEncoder):
(WebGPU::ComputePassEncoder::executePreDispatchCommands):
(WebGPU::ComputePassEncoder::setBindGroup):
(wgpuComputePassEncoderSetBindGroup):
* Source/WebGPU/WebGPU/ComputePipeline.h:
(WebGPU::ComputePipeline::create):
(WebGPU::ComputePipeline::uniqueId const):
* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::ComputePipeline::ComputePipeline):
* Source/WebGPU/WebGPU/Device.h:
(WebGPU::Device::placeholderBufferUniqueId const):
* Source/WebGPU/WebGPU/Pipeline.mm:
(WebGPU::errorValidatingBindGroup):
* Source/WebGPU/WebGPU/PipelineLayout.h:
(WebGPU::PipelineLayout::create):
(WebGPU::PipelineLayout::uniqueId const):
* Source/WebGPU/WebGPU/PipelineLayout.mm:
(WebGPU::Device::createPipelineLayout):
(WebGPU::addInitialOffset):
(WebGPU::PipelineLayout::PipelineLayout):
(WebGPU::returnTotalSize):
(WebGPU::PipelineLayout::offsetVectorForBindGroup):
(WebGPU::PipelineLayout::updateVertexOffsets):
(WebGPU::PipelineLayout::updateFragmentOffsets):
(WebGPU::PipelineLayout::updateComputeOffsets):
(WebGPU::PipelineLayout::vertexOffsets): Deleted.
(WebGPU::PipelineLayout::fragmentOffsets): Deleted.
(WebGPU::PipelineLayout::computeOffsets): Deleted.
* Source/WebGPU/WebGPU/QuerySet.mm:
(WebGPU::Device::createQuerySet):
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::RenderBundleEncoder::executePreDrawCommands):
(WebGPU::RenderBundleEncoder::setBindGroup):
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::makeKey):
(WebGPU::RenderPassEncoder::setVertexBuffer):
(WebGPU::RenderPassEncoder::setFragmentBuffer):
(WebGPU::m_maxDrawCount):
(WebGPU::RenderPassEncoder::RenderPassEncoder):
(WebGPU::RenderPassEncoder::runIndexBufferValidation):
(WebGPU::RenderPassEncoder::runVertexBufferValidation):
(WebGPU::RenderPassEncoder::errorValidatingAndBindingBuffers):
(WebGPU::RenderPassEncoder::setCachedRenderPassState):
(WebGPU::RenderPassEncoder::executePreDrawCommands):
(WebGPU::RenderPassEncoder::computeMininumVertexInstanceCount const):
(WebGPU::RenderPassEncoder::clampIndexBufferToValidValues):
(WebGPU::RenderPassEncoder::clampIndirectIndexBufferToValidValues):
(WebGPU::RenderPassEncoder::clampIndirectBufferToValidValues):
(WebGPU::RenderPassEncoder::splitRenderPass):
(WebGPU::RenderPassEncoder::executeBundles):
(WebGPU::RenderPassEncoder::setBindGroup):
(WebGPU::RenderPassEncoder::setPipeline):
(WebGPU::RenderPassEncoder::setViewport):
(wgpuRenderPassEncoderSetBindGroup):
* Source/WebGPU/WebGPU/RenderPipeline.h:
(WebGPU::RenderPipeline::create):
(WebGPU::RenderPipeline::uniqueId const):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::RenderPipeline::RenderPipeline):
(WebGPU::m_uniqueId):
(WebGPU::RenderPipeline::recomputeLastStrideAsStride const):
(WebGPU::m_minimumBufferSizes): Deleted.
* Source/WebGPU/WebGPU/WebGPU.h:

Canonical link: <a href="https://commits.webkit.org/293608@main">https://commits.webkit.org/293608@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/98a709e91d837e5b1e33e2d260a464b8a6eb0755

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99362 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19011 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9262 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104493 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49963 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101403 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19300 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27445 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75619 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32725 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102369 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14675 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89706 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55978 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14469 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7688 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49323 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84395 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7775 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106850 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26476 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19314 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84579 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26838 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85909 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84091 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21338 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28761 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6450 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20219 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26416 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31616 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26236 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29549 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27803 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/WebCore/Modules/WebGPU/Implementation/WebGPUComputePassEncoderImpl.cpp:
```diff
@@ -71,16 +71,12 @@ void ComputePassEncoderImpl::end()
+void ComputePassEncoderImpl::setBindGroup(Index32, const BindGroup&, std::span<const uint32_t>, Size64, Size32)
 {
-    wgpuComputePassEncoderSetBindGroup(m_backing.get(), index, protectedConvertToBackingContext()->convertToBacking(bindGroup), dynamicOffsetsDataLength, dynamicOffsetsArrayBuffer.subspan(dynamicOffsetsDataStart).data());
+    RELEASE_ASSERT_NOT_REACHED();
```
Why is this needed but never called? Maybe worth a comment?

mwyrzykowski replied:
The reasoning is `ComputePassEncoderImpl` (in GPUP) and `RemoteComputePassEncoderProxy` (in Web content process) both inherit from `ComputePassEncoder` (process agnostic) and only `RemoteComputePassEncoderProxy` supports both member functions as the specification does. But `RemoteComputePassEncoderProxy` translates both member functions into a single member function so the other one will never get called in the GPUP.

It evolved from when WebGPU was designed for WK1 and such a factoring was necessary, but since WebGPU is WK2 only now some of this outdated architecture remains

mwyrzykowski committed:
[WebGPU] GPURenderPassEncoder.drawIndexed takes significant time in GPU process
https://bugs.webkit.org/show_bug.cgi?id=289480
rdar://146676364

Reviewed by Tadeu Zagallo.

executePreDrawCommands was often performing identical validation repeatedely,
leading to to slower framerates than Chrome on some Unity content.

Reduce overhead via caching.

Improves performance on M1 to >60fps in given test cases.

* Source/WebCore/Modules/WebGPU/Implementation/WebGPUComputePassEncoderImpl.cpp:
(WebCore::WebGPU::ComputePassEncoderImpl::setBindGroup):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPURenderPassEncoderImpl.cpp:
(WebCore::WebGPU::RenderPassEncoderImpl::setBindGroup):
* Source/WebGPU/WebGPU/BindGroup.h:
(WebGPU::BindGroup::create):
(WebGPU::BindGroup::vertexArgumentBufferUniqueId const):
(WebGPU::BindGroup::fragmentArgumentBufferUniqueId const):
(WebGPU::BindGroup::computeArgumentBufferUniqueId const):
(WebGPU::BindGroup::uniqueId const):
* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::Device::createBindGroup):
(WebGPU::BindGroup::BindGroup):
(WebGPU::BindGroup::setLabel):
(WebGPU::BindGroup::makeSubmitInvalid const):
(WebGPU::makePipelineBindGroupKey):
(WebGPU::BindGroup::validatedSuccessfully const):
(WebGPU::BindGroup::previouslyValidatedBindGroup const):
* Source/WebGPU/WebGPU/BindGroupLayout.h:
(WebGPU::BindGroupLayout::create):
(WebGPU::BindGroupLayout::uniqueId const):
* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::Device::createBindGroupLayout):
(WebGPU::BindGroupLayout::errorValidatingDynamicOffsets const):
(WebGPU::makeBindGroupLayoutPairIdentifier):
(WebGPU::Device::isCachedCompatibile const):
(WebGPU::Device::setCachedCompatibile const):
(WebGPU::Device::removeCachedBindGroupLayout const):
(WebGPU::BindGroupLayout::errorValidatingBindGroupCompatibility const):
(WebGPU::BindGroupLayout::BindGroupLayout):
(WebGPU::BindGroupLayout::~BindGroupLayout):
* Source/WebGPU/WebGPU/BindableResource.h:
(WebGPU::BufferWithUniqueId::operator id<MTLBuffer> const):
(WebGPU::BufferWithUniqueId::operator uint64_t const):
* Source/WebGPU/WebGPU/Buffer.h:
(WebGPU::Buffer::create):
(WebGPU::Buffer::uniqueId const):
(WebGPU::Buffer::indirectIndexedBufferUniqueId const):
(WebGPU::Buffer::indirectBufferUniqueId const):
* Source/WebGPU/WebGPU/Buffer.mm:
(WebGPU::Device::safeCreateBuffer const):
(WebGPU::Device::createBuffer):
(WebGPU::Buffer::Buffer):
(WebGPU::Buffer::destroy):
(WebGPU::Buffer::getMappedRange):
(WebGPU::Buffer::unmap):
(WebGPU::Buffer::setLabel):
(WebGPU::Buffer::currentSize const):
(WebGPU::Buffer::takeSlowIndexValidationPath):
(WebGPU::Buffer::takeSlowIndirectIndexValidationPath):
(WebGPU::Buffer::takeSlowIndirectValidationPath):
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::encoderIsCurrent const):
* Source/WebGPU/WebGPU/ComputePassEncoder.h:
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::ComputePassEncoder):
(WebGPU::ComputePassEncoder::executePreDispatchCommands):
(WebGPU::ComputePassEncoder::setBindGroup):
(wgpuComputePassEncoderSetBindGroup):
* Source/WebGPU/WebGPU/ComputePipeline.h:
(WebGPU::ComputePipeline::create):
(WebGPU::ComputePipeline::uniqueId const):
* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::ComputePipeline::ComputePipeline):
* Source/WebGPU/WebGPU/Device.h:
(WebGPU::Device::placeholderBufferUniqueId const):
* Source/WebGPU/WebGPU/Pipeline.mm:
(WebGPU::errorValidatingBindGroup):
* Source/WebGPU/WebGPU/PipelineLayout.h:
(WebGPU::PipelineLayout::create):
(WebGPU::PipelineLayout::uniqueId const):
* Source/WebGPU/WebGPU/PipelineLayout.mm:
(WebGPU::Device::createPipelineLayout):
(WebGPU::addInitialOffset):
(WebGPU::PipelineLayout::PipelineLayout):
(WebGPU::returnTotalSize):
(WebGPU::PipelineLayout::offsetVectorForBindGroup):
(WebGPU::PipelineLayout::updateVertexOffsets):
(WebGPU::PipelineLayout::updateFragmentOffsets):
(WebGPU::PipelineLayout::updateComputeOffsets):
(WebGPU::PipelineLayout::vertexOffsets): Deleted.
(WebGPU::PipelineLayout::fragmentOffsets): Deleted.
(WebGPU::PipelineLayout::computeOffsets): Deleted.
* Source/WebGPU/WebGPU/QuerySet.mm:
(WebGPU::Device::createQuerySet):
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::RenderBundleEncoder::executePreDrawCommands):
(WebGPU::RenderBundleEncoder::setBindGroup):
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::makeKey):
(WebGPU::RenderPassEncoder::setVertexBuffer):
(WebGPU::RenderPassEncoder::setFragmentBuffer):
(WebGPU::m_maxDrawCount):
(WebGPU::RenderPassEncoder::RenderPassEncoder):
(WebGPU::RenderPassEncoder::runIndexBufferValidation):
(WebGPU::RenderPassEncoder::runVertexBufferValidation):
(WebGPU::RenderPassEncoder::errorValidatingAndBindingBuffers):
(WebGPU::RenderPassEncoder::setCachedRenderPassState):
(WebGPU::RenderPassEncoder::executePreDrawCommands):
(WebGPU::RenderPassEncoder::computeMininumVertexInstanceCount const):
(WebGPU::RenderPassEncoder::clampIndexBufferToValidValues):
(WebGPU::RenderPassEncoder::clampIndirectIndexBufferToValidValues):
(WebGPU::RenderPassEncoder::clampIndirectBufferToValidValues):
(WebGPU::RenderPassEncoder::splitRenderPass):
(WebGPU::RenderPassEncoder::executeBundles):
(WebGPU::RenderPassEncoder::setBindGroup):
(WebGPU::RenderPassEncoder::setPipeline):
(WebGPU::RenderPassEncoder::setViewport):
(wgpuRenderPassEncoderSetBindGroup):
* Source/WebGPU/WebGPU/RenderPipeline.h:
(WebGPU::RenderPipeline::create):
(WebGPU::RenderPipeline::uniqueId const):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::RenderPipeline::RenderPipeline):
(WebGPU::m_uniqueId):
(WebGPU::RenderPipeline::recomputeLastStrideAsStride const):
(WebGPU::m_minimumBufferSizes): Deleted.
* Source/WebGPU/WebGPU/WebGPU.h:

Canonical link: https://commits.webkit.org/293608@main
